tests/integration/test_completion.py

```diff
@@ -48,18 +49,30 @@ def fill_parser(self, parser: craft_cli.dispatcher._CustomArgumentParser) -> Non
         parser.add_argument("src", type=Path)
         parser.add_argument("dest", type=Path)
 
-def test_completion() -> None:
-    def _get_dispatcher() -> craft_cli.Dispatcher:
-        basic_group = craft_cli.CommandGroup("basic", [FakeLsCommand, FakeCpCommand])
+def get_app_info_func(commands: Sequence[Type[craft_cli.BaseCommand]], config: Dict[str, Any] = {}) -> Callable[[], Tuple[craft_cli.Dispatcher, Dict[str, Any]]]:
+    basic_group = craft_cli.CommandGroup("basic", commands)
 
+    def _inner() -> Tuple[craft_cli.Dispatcher, Dict[str, Any]]:
         return craft_cli.Dispatcher(
             appname="pybash",
             commands_groups=[basic_group],
             extra_global_args=[],
-        )
+        ), config
+
+    return _inner
 
-    actual_output = complete("testcraft", _get_dispatcher)
+def test_completion_output() -> None:
+    app_info_func = get_app_info_func([FakeLsCommand, FakeCpCommand])
+    actual_output = complete("testcraft", app_info_func)
 
     expected_output = (Path(__file__).parent / "test_completion" / "expected_script.sh").read_text()
 
     assert actual_output == expected_output
+
+def test_app_config_used() -> None:
+    app_info_func = get_app_info_func([FakeCpCommand], config={"hello": "world"})
+
+    with patch(__name__ + ".FakeCpCommand.__init__", return_value=None) as complete_mock:
```

Comment:
I'm not a big fan of mocking inits if we don't have to, it makes the tests harder to reason about.

How about this instead: Create a separate command that, in its init, adds a key to that `config` dict. Then you can check that the dict has the key after calling `complete`.